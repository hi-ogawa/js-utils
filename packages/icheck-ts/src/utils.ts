import fs from "node:fs";
import path from "node:path";
import { LruCache, hashString, memoize } from "@hiogawa/utils";

export function memoizeOnFile<F extends (...args: any[]) => any>(
  f: F,
  options: { disabled?: boolean; file: string; maxSize: number }
) {
  if (options.disabled) {
    return [f, undefined] as const;
  }

  const cache = new LruCache<string, any>(options.maxSize);

  async function load() {
    if (fs.existsSync(options.file)) {
      const raw = await fs.promises.readFile(options.file, "utf-8");
      const data = JSON.parse(raw);
      cache._map = new Map(data);
    }
  }

  async function save() {
    const data = JSON.stringify([...cache._map.entries()]);
    await fs.promises.mkdir(path.dirname(options.file), { recursive: true });
    await fs.promises.writeFile(options.file, data);
  }

  const memoized = memoize(f, {
    keyFn: (...args) => hashString(JSON.stringify(args)),
    cache,
  });

  return [memoized, { load, save }] as const;
}
