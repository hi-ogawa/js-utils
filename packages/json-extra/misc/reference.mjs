/*
usage:
  node ./packages/json-extra/misc/reference.mjs input console json-extra devalue
*/

// INPUT_START
const parent = { children: new Map() };
const child1 = { parent };
const child2 = { parent, siblings: new Set([child1]) };
parent.children.set("foo", child1);
parent.children.set("bar", child2);
// INPUT_END

const args = process.argv.slice(2);
const input = parent;

if (args.includes("input")) {
  const fs = await import("node:fs");
  const data = fs.readFileSync(new URL(import.meta.url), "utf-8");
  const re = new RegExp("// INPUT_START\n(.*)\n// INPUT_END", "s");
  console.log(data.match(re)[1]);
}

if (args.includes("console")) {
  console.log(input);
}

if (args.includes("json-extra")) {
  const { createJsonExtra } = await import("../dist/reference.js");
  const jsonExtra = createJsonExtra({ builtins: true });
  const stringfied = jsonExtra.stringify(input, null, 2);
  console.log(stringfied);
}

if (args.includes("devalue")) {
  // https://github.com/Rich-Harris/devalue
  const devalue = await import("devalue");
  const stringified = devalue.stringify(input);
  console.log(JSON.stringify(JSON.parse(stringified), null, 2));
}
