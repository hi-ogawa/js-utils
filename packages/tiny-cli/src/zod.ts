import type { z } from "zod";
import type { ArgSchema } from "./typed";

// helper to reuse zod schema by monkey-patching...

const ARG_META = Symbol("ARG_META");

// cf. https://github.com/anatine/zod-plugins/blob/547e9a89e14ca889367b80791e8cc15393cba458/packages/zod-openapi/src/lib/zod-extensions.ts#L11C1-L19C2
declare module "zod" {
  interface ZodSchema<
    Output = any,
    Def extends z.ZodTypeDef = z.ZodTypeDef,
    Input = Output
  > {
    [ARG_META]: Omit<ArgSchema<unknown>, "parse">;

    asArg<T extends ZodSchema<Output, Def, Input>>(
      this: T,
      argMeta: Omit<ArgSchema<unknown>, "parse">
    ): T;
  }
}

export function setupZodExtendArg(zod: typeof z) {
  zod.ZodSchema.prototype.asArg = function (this, argMeta) {
    this[ARG_META] = argMeta;
    return this;
  };
  return () => {
    delete (zod.ZodSchema.prototype as any).extendArg;
  };
}

export function zodArg<Schema extends z.ZodType>(
  schema: Schema
): ArgSchema<z.infer<Schema>> {
  return {
    parse: schema.parse,
    help: schema.description,
    ...schema[ARG_META],
  };
}

export function zodArgs<Schema extends z.SomeZodObject, T = z.infer<Schema>>(
  schema: Schema
): { [K in keyof T]: ArgSchema<T[K]> } {
  let result: any = {};
  for (const [k, v] of Object.entries(schema.shape)) {
    result[k] = zodArg(v);
  }
  return result;
}
