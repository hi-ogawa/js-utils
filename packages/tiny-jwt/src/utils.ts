export function encodeJson(v: unknown) {
  return encodeBase64url(encodeUtf8(JSON.stringify(v)));
}

export function decodeJson(v: string): unknown {
  return JSON.parse(decodeUtf8(decodeBase64url(v)));
}

//
// string <-> buffer
//

export const encodeUtf8 = (v: string) => new TextEncoder().encode(v);
export const decodeUtf8 = (v: BufferSource) => new TextDecoder().decode(v);

//
// base64url string <-> buffer
// https://developer.mozilla.org/en-US/docs/Glossary/Base64#the_unicode_problem)
// https://datatracker.ietf.org/doc/html/rfc7515#appendix-C
//

export function encodeBase64url(buffer: Uint8Array): string {
  const binString = Array.from(buffer, (c) => String.fromCharCode(c)).join("");
  const base64 = btoa(binString);
  const base64url = base64
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
  return base64url;
}

export function decodeBase64url(base64url: string): Uint8Array {
  // `atob` can handle without padding
  const base64 = base64url.replace(/-/g, "+").replace(/_/g, "/");
  const binString = atob(base64);
  const buffer = Uint8Array.from(binString, (c) => c.charCodeAt(0)!);
  return buffer;
}
