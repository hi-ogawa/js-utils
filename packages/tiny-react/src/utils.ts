export function isEqualShallow(xs: unknown[], ys: unknown[]) {
  return xs.length === ys.length && xs.every((x, i) => x === ys[i]);
}
