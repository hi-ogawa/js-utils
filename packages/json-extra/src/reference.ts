import { tinyassert } from "@hiogawa/utils";

export function replaceReference(data: unknown) {
  const refs = new Map<unknown, number>();

  // TODO: configurable
  const plugins: Record<string, Plugin<any>> = builtinPlugins;

  function recurse(v: unknown) {
    // return reference placeholder
    if (refs.has(v)) {
      return ["!", refs.get(v)];
    }

    // track reference
    if (v && typeof v === "object") {
      refs.set(v, refs.size);
    }

    // escape custom encoding collision
    //   ["!xxx", ...] ==> ["!", "!xxx", ....]
    if (
      Array.isArray(v) &&
      v.length >= 2 &&
      typeof v[0] === "string" &&
      v[0][0] === "!"
    ) {
      v = ["!", ...v];
    }

    // custom replacer
    for (const tag in plugins) {
      const p = plugins[tag];
      if (p.is(v)) {
        v = [`!${tag}`, p.replacer(v)];
        break;
      }
    }

    if (v && typeof v === "object") {
      v = shallowCopy(v);
      for (const [k, e] of Object.entries(v as any)) {
        (v as any)[k] = recurse(e);
      }
    }
    return v;
  }

  return recurse(data);
}

export function reviveReference(data: unknown) {
  const refs: unknown[] = [];

  // TODO: configurable
  const plugins: Record<string, Plugin<any>> = builtinPlugins;

  function recurse(v: unknown) {
    // revive reference placeholder
    if (Array.isArray(v) && v.length === 2 && v[0] === "!") {
      return refs[v[1]];
    }

    let plugin: Plugin<any> | undefined;
    let ref: unknown | undefined;

    // TODO: refactor messy branches...

    // unescape custom encoding collision
    //   ["!xxx", ...] <== ["!", "!xxx", ....]
    if (Array.isArray(v) && v.length >= 3 && v[0] === "!") {
      v = v.slice(1);
      refs.push(v);
    } else if (
      // custom reviver to obtain ref before recurse
      Array.isArray(v) &&
      v.length === 2 &&
      typeof v[0] === "string" &&
      v[0][0] === "!"
    ) {
      const tag = v[0].slice(1);
      v = shallowCopy(v[1]);
      plugin = plugins[tag];
      tinyassert(plugin);
      ref = plugin.reviverRef();
      refs.push(ref);
      refs.push("__dummy"); // dummy ref to align with replaceReference's offset (TODO: probably should fix replaceReference side)
    } else if (v && typeof v === "object") {
      // track ref before recurse to handle circular references
      v = shallowCopy(v);
      refs.push(v);
    }

    if (v && typeof v === "object") {
      for (const [k, e] of Object.entries(v as any)) {
        (v as any)[k] = recurse(e);
      }
    }

    // custom reviver after recurse
    if (plugin) {
      v = plugin.reviver(v, ref);
    }

    return v;
  }

  return recurse(data);
}

// shallow copy to avoid overwriting original input
function shallowCopy(v: unknown) {
  if (v && typeof v === "object") {
    return Array.isArray(v) ? [...v] : { ...v };
  }
  return v;
}

//
// plugin
//

type Plugin<T> = {
  is: (v: unknown) => boolean;
  replacer: (v: T) => unknown;
  reviver: (v: unknown, ref: T) => T;
  reviverRef: () => T;
};

// type helper
export function definePlugin<T>(v: Plugin<T>): Plugin<T> {
  return v;
}

function definePluginConstant<T>(c: T): Plugin<T> {
  return {
    is: (v) => Object.is(v, c),
    replacer: () => 0,
    reviver: () => c,
    reviverRef: () => c,
  };
}

export const builtinPlugins = {
  //
  // constants
  //
  undefined: definePluginConstant(undefined),
  Infinity: definePluginConstant(Infinity),
  "-Infinity": definePluginConstant(-Infinity),
  NaN: definePluginConstant(NaN),
  "-0": definePluginConstant(-0),

  //
  // non containers
  //
  Date: definePlugin<Date>({
    is: (v) => v instanceof Date,
    replacer: (v) => v.toISOString(),
    reviver: (v, ref) => {
      tinyassert(typeof v === "string");
      ref.setTime(new Date(v).getTime());
      return ref;
    },
    reviverRef: () => new Date(0),
  }),

  //
  // containers
  //
  Set: definePlugin<Set<unknown>>({
    is: (v) => v instanceof Set,
    replacer: (v) => Array.from(v),
    reviver: (v, ref) => {
      tinyassert(Array.isArray(v));
      for (const e of v) {
        ref.add(e);
      }
      return ref;
    },
    reviverRef: () => new Set(),
  }),
};
