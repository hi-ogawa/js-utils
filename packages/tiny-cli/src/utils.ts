import { flattenErrorCauses, range } from "@hiogawa/utils";

export const DEFAULT_PROGRAM = "(cli)";

export class ParseError extends Error {
  static wrapFn<T>(message: string, f: () => T): T {
    try {
      return f();
    } catch (e) {
      throw new ParseError(message, { cause: e });
    }
  }
}

export function formatTable(rows: string[][]) {
  return formatIndent(
    padColumns(rows).map((row) => row.join(" ".repeat(4)).trimEnd()),
    2
  );
}

function padColumns(rows: string[][]): string[][] {
  if (rows.length === 0) {
    return rows;
  }
  const ncol = Math.max(...rows.map((row) => row.length));
  const widths = range(ncol).map((c) =>
    Math.max(...rows.map((row) => row[c]?.length ?? 0))
  );
  const newRows = rows.map((row) =>
    row.map((v, i) => v.padEnd(widths[i], " "))
  );
  return newRows;
}

function formatIndent(ls: string[], n: number): string {
  return ls.map((v) => " ".repeat(n) + v).join("\n");
}

// consola.error + flattenErrorCauses
// to provide good-enough cli error handling experience by default
// TODO: maybe move it to `packages/utlis`?
export function consoleErrorExtra(e: unknown) {
  for (const err of flattenErrorCauses(e)) {
    if (err instanceof Error) {
      console.error(formatError(err));
    } else {
      console.error(err);
    }
  }
}

// simpler port of consola.error
// https://github.com/unjs/consola/blob/e4a37c1cd2c8d96b5f30d8c13ff2df32244baa6a/src/utils/color.ts#L93
// https://github.com/unjs/consola/blob/e4a37c1cd2c8d96b5f30d8c13ff2df32244baa6a/src/reporters/fancy.ts#L49
// https://github.com/unjs/consola/blob/e4a37c1cd2c8d96b5f30d8c13ff2df32244baa6a/src/reporters/fancy.ts#L64-L65
function formatError(e: Error) {
  // color?
  // \u001B[41m ERROR \u001B[49m
  // \u001B[36m${stack}\u001B[39m

  let result = `\

[ERROR] ${e.message}

`;
  if (e.stack) {
    const stack = e.stack.split("\n").slice(1).join("\n");
    result += `\
${stack}
`;
  }
  return result;
}
