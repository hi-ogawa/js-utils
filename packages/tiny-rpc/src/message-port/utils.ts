import { DefaultMap } from "@hiogawa/utils";

export class TypedEventTarget<T> {
  private listeners = new DefaultMap<keyof T, Set<Function>>(() => new Set());

  addEventListener<K extends keyof T>(type: K, listener: (ev: T[K]) => void) {
    this.listeners.get(type).add(listener);
  }

  removeEventListener<K extends keyof T>(
    type: K,
    listener: (ev: T[K]) => void,
  ) {
    this.listeners.get(type).delete(listener);
  }

  notify<K extends keyof T>(type: K, data: T[K]) {
    for (const listener of this.listeners.get(type)) {
      listener(data);
    }
  }
}

export function subscribe<K extends keyof EventSourceEventMap>(
  target: EventSource,
  type: K,
  listener: (ev: EventSourceEventMap[K]) => void,
) {
  target.addEventListener(type, listener);
  return () => {
    target.addEventListener(type, listener);
  };
}
