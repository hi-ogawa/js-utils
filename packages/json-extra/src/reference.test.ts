import { tinyassert } from "@hiogawa/utils";
import fc from "fast-check";
import { describe, expect, it } from "vitest";
import { ZodError, z } from "zod";
import { createJsonExtra2, definePlugin } from "./reference";

// instantiate default one for tests
const jsonExtra = createJsonExtra2({ builtins: true });

function runJsonExtra(input: unknown, json = jsonExtra) {
  const stringified = json.stringify(input, null, 2);
  const parsed = json.parse(stringified);
  const serialized = json.serialize(input);
  const deserialized = json.deserialize(serialized);
  return { input, stringified, parsed, serialized, deserialized };
}

describe("reference", () => {
  it("basic", () => {
    const v1 = { hi: 0 };
    const v = [v1, [v1], { foo: v1 }];
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        {
          "hi": 0,
        },
        [
          [
            "!",
            1,
          ],
        ],
        {
          "foo": [
            "!",
            1,
          ],
        },
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        {
          "hi": 0,
        },
        [
          {
            "hi": 0,
          },
        ],
        {
          "foo": {
            "hi": 0,
          },
        },
      ]
    `);
    expect(u).toEqual(v);
    expect(u[0]).toBe(u[1][0]);
    expect(u[0]).toBe(u[2].foo);
  });

  it("two", () => {
    const v1 = { hi: 0 };
    const v2 = { hey: 1, yo: v1 };
    const v = [v1, [v1], { ...v1 }, v2, { foo: v1, bar: [v2] }];
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        {
          "hi": 0,
        },
        [
          [
            "!",
            1,
          ],
        ],
        {
          "hi": 0,
        },
        {
          "hey": 1,
          "yo": [
            "!",
            1,
          ],
        },
        {
          "bar": [
            [
              "!",
              4,
            ],
          ],
          "foo": [
            "!",
            1,
          ],
        },
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        {
          "hi": 0,
        },
        [
          {
            "hi": 0,
          },
        ],
        {
          "hi": 0,
        },
        {
          "hey": 1,
          "yo": {
            "hi": 0,
          },
        },
        {
          "bar": [
            {
              "hey": 1,
              "yo": {
                "hi": 0,
              },
            },
          ],
          "foo": {
            "hi": 0,
          },
        },
      ]
    `);
    expect(u).toEqual(v);
    expect(u[0]).toBe(u[1][0]);
    expect(u[0]).not.toBe(u[2]);
    expect(u[3]).toBe(u[4].bar[0]);
  });

  it("cyclic", () => {
    const v: any = [];
    v[0] = v;
    v[1] = { hi: v };
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        [
          "!",
          0,
        ],
        {
          "hi": [
            "!",
            0,
          ],
        },
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        [Circular],
        {
          "hi": [Circular],
        },
      ]
    `);
    expect(u).toBe(u[0]);
    expect(u).toBe(u[1].hi);
  });

  it("collision 1", () => {
    const v1 = { hi: 0 };
    const v2 = ["!", v1];
    const v = [v1, v2, v2];
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        {
          "hi": 0,
        },
        [
          "!",
          "!",
          [
            "!",
            1,
          ],
        ],
        [
          "!",
          2,
        ],
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        {
          "hi": 0,
        },
        [
          "!",
          {
            "hi": 0,
          },
        ],
        [
          "!",
          {
            "hi": 0,
          },
        ],
      ]
    `);
    expect(u).toEqual(v);
    expect(u[0]).toBe(u[1][1]);
    expect(u[1]).toBe(u[2]);
  });

  it("collision 2", () => {
    const v1 = { hi: 0 };
    const v2 = ["!", v1];
    const v = [
      v1,
      [v1],
      ["!"],
      v2,
      ["!", { "!": v1 }],
      ["!!", "!", "!"],
      ["!", v2],
    ];
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        {
          "hi": 0,
        },
        [
          [
            "!",
            1,
          ],
        ],
        [
          "!",
        ],
        [
          "!",
          "!",
          [
            "!",
            1,
          ],
        ],
        [
          "!",
          "!",
          {
            "!": [
              "!",
              1,
            ],
          },
        ],
        [
          "!",
          "!!",
          "!",
          "!",
        ],
        [
          "!",
          "!",
          [
            "!",
            4,
          ],
        ],
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        {
          "hi": 0,
        },
        [
          {
            "hi": 0,
          },
        ],
        [
          "!",
        ],
        [
          "!",
          {
            "hi": 0,
          },
        ],
        [
          "!",
          {
            "!": {
              "hi": 0,
            },
          },
        ],
        [
          "!!",
          "!",
          "!",
        ],
        [
          "!",
          [
            "!",
            {
              "hi": 0,
            },
          ],
        ],
      ]
    `);
    expect(u).toEqual(v);
    expect(u[0]).toBe(u[1][0]);
    expect(u[0]).toBe(u[3][1]);
    expect(u[0]).toBe(u[4][1]["!"]);
  });

  it("custom single", () => {
    const v1 = new Date("2000-01-01T00:00:00Z");
    const v2 = /^\d+/gms;
    const v3 = [v1, v2];
    const v = [v1, [v1], { foo: [v1, v2] }, v2, v3, v3];
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        [
          "!Date",
          "2000-01-01T00:00:00.000Z",
        ],
        [
          [
            "!",
            1,
          ],
        ],
        {
          "foo": [
            [
              "!",
              1,
            ],
            [
              "!RegExp",
              [
                "^\\d+",
                "gms",
              ],
            ],
          ],
        },
        [
          "!",
          5,
        ],
        [
          [
            "!",
            1,
          ],
          [
            "!",
            5,
          ],
        ],
        [
          "!",
          6,
        ],
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        2000-01-01T00:00:00.000Z,
        [
          2000-01-01T00:00:00.000Z,
        ],
        {
          "foo": [
            2000-01-01T00:00:00.000Z,
            /\\^\\\\d\\+/gms,
          ],
        },
        /\\^\\\\d\\+/gms,
        [
          2000-01-01T00:00:00.000Z,
          /\\^\\\\d\\+/gms,
        ],
        [
          2000-01-01T00:00:00.000Z,
          /\\^\\\\d\\+/gms,
        ],
      ]
    `);
    expect(u).toEqual(v);
    expect(u[0]).toBe(u[1][0]);
    expect(u[0]).toBe(u[2].foo[0]);
    expect(u[3]).toBe(u[2].foo[1]);
  });

  it("custom constant", () => {
    const v1 = [undefined, Infinity, NaN];
    const v = [v1, [v1], undefined, new Date(1), Infinity, new Date(0)];
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        [
          [
            "!undefined",
            0,
          ],
          [
            "!Infinity",
            0,
          ],
          [
            "!NaN",
            0,
          ],
        ],
        [
          [
            "!",
            1,
          ],
        ],
        [
          "!undefined",
          0,
        ],
        [
          "!Date",
          "1970-01-01T00:00:00.001Z",
        ],
        [
          "!Infinity",
          0,
        ],
        [
          "!Date",
          "1970-01-01T00:00:00.000Z",
        ],
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        [
          undefined,
          Infinity,
          NaN,
        ],
        [
          [
            undefined,
            Infinity,
            NaN,
          ],
        ],
        undefined,
        1970-01-01T00:00:00.001Z,
        Infinity,
        1970-01-01T00:00:00.000Z,
      ]
    `);
    expect(u).toEqual(v);
    expect(u[0]).toBe(u[1][0]);
  });

  it("custom container 1", () => {
    const v1 = new Date("2000-01-01T00:00:00Z");
    const v2 = new Set([v1, { foo: v1 }]);
    const v = [v2, [v2]];
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        [
          "!Set",
          [
            [
              "!Date",
              "2000-01-01T00:00:00.000Z",
            ],
            {
              "foo": [
                "!",
                2,
              ],
            },
          ],
        ],
        [
          [
            "!",
            1,
          ],
        ],
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        Set {
          2000-01-01T00:00:00.000Z,
          {
            "foo": 2000-01-01T00:00:00.000Z,
          },
        },
        [
          Set {
            2000-01-01T00:00:00.000Z,
            {
              "foo": 2000-01-01T00:00:00.000Z,
            },
          },
        ],
      ]
    `);
    expect(u).toEqual(v);
    expect(u[0]).toBe(u[1][0]);
    const [x, y] = [...u[0]];
    expect(x).toBe(y.foo);
  });

  it("custom container cyclic", () => {
    const v1 = new Set(["hi" as any]);
    v1.add(v1);
    const v = [v1, [v1]];
    const serialized = jsonExtra.serialize(v);
    expect(serialized).toMatchInlineSnapshot(`
      [
        [
          "!Set",
          [
            "hi",
            [
              "!",
              1,
            ],
          ],
        ],
        [
          [
            "!",
            1,
          ],
        ],
      ]
    `);
    const u = jsonExtra.deserialize(serialized);
    expect(u).toMatchInlineSnapshot(`
      [
        Set {
          "hi",
          [Circular],
        },
        [
          Set {
            "hi",
            [Circular],
          },
        ],
      ]
    `);
    expect(u).toEqual(v);
    expect(u[0]).toBe(u[1][0]);
    expect(u[0]).toBe([...u[0]][1]);
  });

  it("edge cases 1", () => {
    const v = new Set(["!", null]);
    const result = runJsonExtra(v);
    expect(result.serialized).toMatchInlineSnapshot(`
      [
        "!Set",
        [
          "!",
          null,
        ],
      ]
    `);
    expect(result.deserialized).toMatchInlineSnapshot(`
      Set {
        "!",
        null,
      }
    `);
    expect(result.deserialized).toEqual(v);
  });

  it("edge cases 2", () => {
    const v = new Set(["!x", null]);
    const result = runJsonExtra(v);
    expect(result.serialized).toMatchInlineSnapshot(`
      [
        "!Set",
        [
          "!x",
          null,
        ],
      ]
    `);
    expect(result.deserialized).toMatchInlineSnapshot(`
      Set {
        "!x",
        null,
      }
    `);
    expect(result.deserialized).toEqual(v);
  });
});

describe("fuzzing", () => {
  it("jsonValue", () => {
    fc.assert(
      fc.property(fc.jsonValue(), (data) => {
        const result = runJsonExtra(data);
        expect(result.parsed).toEqual(data);
        expect(result.deserialized).toEqual(data);
      })
    );
  });

  it("anything", () => {
    fc.assert(
      fc.property(
        fc.anything({
          withBigInt: true,
          withDate: true,
          withMap: true,
          withSet: true,
        }),
        (data) => {
          const result = runJsonExtra(data);
          expect(result.parsed).toEqual(data);
          expect(result.deserialized).toEqual(data);
        }
      )
    );
  });
});

describe("old tests", () => {
  it("basic", () => {
    const input = [
      // standard json value
      null,
      true,
      123,
      "string",
      ["array"],
      { k: "v" },
      // special constants
      undefined,
      Infinity,
      -Infinity,
      NaN,
      0,
      -0,
      // extra types
      new Date("2023-08-17"),
      1234n,
      /^\d+/gms,
      // extra containers
      new Map<any, any>([
        [0, new Date(0)],
        [1n, new Set([/a/g])],
      ]),
      new Set<any>([0, new Date(0), new Map([[1n, /a/g]])]),
      // escape
      ["!NaN", "collision"],
    ];
    const result = runJsonExtra(input);
    expect(result.stringified).toMatchSnapshot();
    expect(result.parsed).toEqual(input);
  });

  it("custom type", () => {
    const jsonExtra = createJsonExtra2({
      plugins: {
        ZodError: definePlugin<ZodError>({
          type: "simple",
          is: (v): v is ZodError => v instanceof ZodError,
          replace: (v) => v.issues,
          revive: (s) => new ZodError(s as any),
        }),
      },
    });

    const error = z.object({ int: z.number().int() }).safeParse({ int: 1.23 });
    tinyassert(!error.success);

    const input = {
      ok: false,
      value: error.error,
    };
    const result = runJsonExtra(input, jsonExtra);
    expect(result.serialized).toMatchInlineSnapshot(`
      {
        "ok": false,
        "value": [
          "!ZodError",
          [
            {
              "code": "invalid_type",
              "expected": "integer",
              "message": "Expected integer, received float",
              "path": [
                "int",
              ],
              "received": "float",
            },
          ],
        ],
      }
    `);
    expect(result.deserialized).toEqual(input);
  });

  it("selected builtins", () => {
    const jsonExtra = createJsonExtra2({ builtins: ["undefined", "Date"] });
    const input = [undefined, new Date("2023-08-17"), NaN, new Set([0, 1])];
    const result = runJsonExtra(input, jsonExtra);
    expect(result.stringified).toMatchInlineSnapshot(`
      "[
        [
          "!undefined",
          0
        ],
        [
          "!Date",
          "2023-08-17T00:00:00.000Z"
        ],
        null,
        {}
      ]"
    `);
    expect(result.parsed).toMatchInlineSnapshot(`
      [
        undefined,
        2023-08-17T00:00:00.000Z,
        null,
        {},
      ]
    `);
  });

  it("escape collision", () => {
    const input = {
      collision2: ["!", 1n],
      collision3: ["!", ["!", 0]],
      collision4: ["!", ["!", 1n, "!"]],
      collision5: [[], ["!"], ["!", 0], ["!", 0, 0], ["!", 0, 0, 0]],
    };
    const result = runJsonExtra(input);
    expect(result.serialized).toMatchSnapshot();
    expect(result.deserialized).toEqual(input);
  });

  it("edge cases", () => {
    // edge cases which don't roundtrip
    // - function
    // - custom class
    // - custom toJSON

    class X {
      name = "hello";
    }

    class Y {
      toJSON() {
        return "foo";
      }
    }

    const input = [
      Symbol("unique"),
      Symbol.for("named"),
      new X(),
      new Y(),
      {
        toJSON: () => "zzz",
      },
      {
        toJSON: "www",
      },
      {
        toJSON: () => () => "uuu",
      },
      () => {},
    ];

    const result = runJsonExtra(input);
    expect(result).toMatchSnapshot();
  });
});
