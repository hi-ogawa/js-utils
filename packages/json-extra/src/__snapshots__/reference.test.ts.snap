// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`old tests > basic 1`] = `
"[
  null,
  true,
  123,
  "string",
  [
    "array"
  ],
  {
    "k": "v"
  },
  [
    "!undefined",
    0
  ],
  [
    "!Infinity",
    0
  ],
  [
    "!-Infinity",
    0
  ],
  [
    "!NaN",
    0
  ],
  0,
  [
    "!-0",
    0
  ],
  [
    "!Date",
    "2023-08-17T00:00:00.000Z"
  ],
  [
    "!BigInt",
    "1234"
  ],
  [
    "!RegExp",
    [
      "^\\\\d+",
      "gms"
    ]
  ],
  [
    "!Map",
    [
      [
        0,
        [
          "!Date",
          "1970-01-01T00:00:00.000Z"
        ]
      ],
      [
        [
          "!BigInt",
          "1"
        ],
        [
          "!Set",
          [
            [
              "!RegExp",
              [
                "a",
                "g"
              ]
            ]
          ]
        ]
      ]
    ]
  ],
  [
    "!Set",
    [
      0,
      [
        "!Date",
        "1970-01-01T00:00:00.000Z"
      ],
      [
        "!Map",
        [
          [
            [
              "!BigInt",
              "1"
            ],
            [
              "!RegExp",
              [
                "a",
                "g"
              ]
            ]
          ]
        ]
      ]
    ]
  ],
  [
    "!",
    "!NaN",
    "collision"
  ]
]"
`;

exports[`old tests > edge cases 1`] = `
{
  "deserialized": [
    Symbol(unique),
    Symbol(named),
    {
      "name": "hello",
    },
    {},
    "zzz",
    {
      "toJSON": "www",
    },
    [Function],
    [Function],
  ],
  "input": [
    Symbol(unique),
    Symbol(named),
    X {
      "name": "hello",
    },
    "foo",
    "zzz",
    {
      "toJSON": "www",
    },
    [Function],
    [Function],
  ],
  "parsed": [
    null,
    null,
    {
      "name": "hello",
    },
    {},
    "zzz",
    {
      "toJSON": "www",
    },
    null,
    null,
  ],
  "serialized": [
    Symbol(unique),
    Symbol(named),
    {
      "name": "hello",
    },
    {},
    "zzz",
    {
      "toJSON": "www",
    },
    [Function],
    [Function],
  ],
  "stringified": "[
  null,
  null,
  {
    "name": "hello"
  },
  {},
  "zzz",
  {
    "toJSON": "www"
  },
  null,
  null
]",
}
`;

exports[`old tests > escape collision 1`] = `
{
  "collision2": [
    "!",
    "!",
    [
      "!BigInt",
      "1",
    ],
  ],
  "collision3": [
    "!",
    "!",
    [
      "!",
      "!",
      0,
    ],
  ],
  "collision4": [
    "!",
    "!",
    [
      "!",
      "!",
      [
        "!BigInt",
        "1",
      ],
      "!",
    ],
  ],
  "collision5": [
    [],
    [
      "!",
    ],
    [
      "!",
      "!",
      0,
    ],
    [
      "!",
      "!",
      0,
      0,
    ],
    [
      "!",
      "!",
      0,
      0,
      0,
    ],
  ],
}
`;
